src/main/java/net/dries007/tfc/common/tileentity/PotTileEntity.java

```diff
@@ -0,0 +1,199 @@
+package net.dries007.tfc.common.tileentity;
+
+import javax.annotation.Nullable;
+
+import net.minecraft.block.BlockState;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.PlayerInventory;
+import net.minecraft.inventory.container.Container;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.tileentity.TileEntityType;
+import net.minecraft.util.Direction;
+import net.minecraft.util.text.ITextComponent;
+import net.minecraft.util.text.TranslationTextComponent;
+import net.minecraftforge.common.capabilities.Capability;
+import net.minecraftforge.common.util.LazyOptional;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.capability.CapabilityFluidHandler;
+import net.minecraftforge.fluids.capability.IFluidHandler;
+import net.minecraftforge.fluids.capability.templates.FluidTank;
+import net.minecraftforge.items.ItemStackHandler;
+
+import net.dries007.tfc.common.container.PotContainer;
+import net.dries007.tfc.common.recipes.FluidInventoryRecipeWrapper;
+import net.dries007.tfc.common.recipes.IPotRecipe;
+import net.dries007.tfc.common.recipes.TFCRecipeTypes;
+import net.dries007.tfc.common.types.FuelManager;
+
+import static net.dries007.tfc.TerraFirmaCraft.MOD_ID;
+
+public class PotTileEntity extends FirepitTileEntity
+{
+    private static final ITextComponent NAME = new TranslationTextComponent(MOD_ID + ".tile_entity.pot");
+
+    public static final int SLOT_EXTRA_INPUT_START = 4;
+    public static final int SLOT_EXTRA_INPUT_END = 8;
+
+    private IPotRecipe.Output output;
+    private IPotRecipe cachedPotRecipe;
+    private int boilingTicks;
+    protected FluidTank tank;
+    protected final LazyOptional<IFluidHandler> fluidCapability;
```

Comment:
Like the small vessel in 1.12, imo the best way to approach this would be to have a single joint capability implementation which holds both fluids and items. It gets rid of the need for multiple capabilities (as we can cast to `PotTileEntity.IPotCapability` with `.filter(...).map(...)` in `LazyOptional` declarations), and removes the need for `FluidInventoryRecipeWrapper` entirely - you just pass in the pot capability interface, within a `RecipeWrapper` subclass that's generic on the interface type. That makes the wrappers simpler, as the recipe itself becomes `PotRecipe extends SimpleRecipe<PotCapability>`